### PR TITLE
Add `AlphaHoverCard` component

### LINKED ISSUE #8931: [EPIC] Implement new `HoverCard` on Orders index table in the admin
1. Look into what Chloe has shipped so far
2. Read context below
3. Scope out work to actually ship the action tooltip to web
4. Ship

=== context ===

**Prototype:** https://tooltip-prototype.shopifycloud.com/6/action
**Figma:** https://www.figma.com/file/64twmvaeh0gEXs3pZcueRe/Tooltips?node-id=461-11097&viewport=-417%2C318%2C0.15
**Parent issue:** https://github.com/Shopify/web/issues/78351. 
**Raised by:** @Shopify/core-operate-workflows-quality-crew 

```[tasklist]
### Remaining Tasks
- [ ] https://github.com/Shopify/polaris/issues/11743
- [ ] https://github.com/Shopify/polaris/issues/11744
- [ ] https://github.com/Shopify/polaris/issues/11746
- [ ] https://github.com/Shopify/polaris/issues/11769
- [ ] https://github.com/Shopify/polaris/issues/11748
- [ ] https://github.com/Shopify/polaris/issues/11749
- [ ] https://github.com/Shopify/polaris/issues/11750
- [ ] https://github.com/Shopify/polaris/issues/11737
- [x] Revert Popoover integration in the IndexTableCell implementation
- [ ] Find a way to port keyboard accessbility changes from Popoover implementation to base implementation
- [x] Add a debounce to the onMouseEnterCell event listener (so there's a delay in HoverCard activation on hover for perf. reasons)
- [x] https://github.com/Shopify/polaris/issues/11768
- [ ] Use this snapshot for prototype `yarn add @shopify/polaris@0.0.0-snapshot-20240319162906`
```

```[tasklist]
### Pending decisions
- [ ] Who owns the HoverCard going forward?
- [ ] https://github.com/Shopify/polaris/issues/11742
```
```[tasklist]
### Decisions
- [ ] Retain hover behaviour for the HoverCard activation in IndexTable.**
```
** Caveat:
If we require a delay for queries, it needs to be consistent across all hover cards. We can't have some that open instantly, some that are delayed
Demo the above and confirm alignment

### PR DESCRIPTION
**Pull request by chloerice**
### WHY are these changes introduced?

Part 1 of #8931 

### WHAT is this pull request doing?

This PR adds a new `AlphaHoverCard` component, which will be used for hover-only detail previews on commerce object links across the admin.

[Storybook](https://5d559397bae39100201eedc1-kktwqzknsy.chromatic.com/?path=/story/all-components-hovercard--with-dynamic-activator) 📖 

Demo

https://github.com/Shopify/polaris/assets/18447883/c11196b7-165b-493b-bfae-983466e4b08e

### To do
- [ ] Bug fixes
  - [ ] Fix child activated hover card not opening when preferred position is bottom
  - [ ] Fix PositionedOverlay render out of bounds initially when height of contents increases beyond available height, causing temporary scrollbar (if there isn't one) that messes up horizontal position and ruins animation (not reproducible in the admin
    - Also think this is what is causing children activated hover card not to render when preferred position is vertical and calculateVerticalPosition.positioning === “below”
- [x] /orders implementation ([PR](https://github.com/Shopify/web/pull/116631))
- [x] Test coverage
- [x] Responsive left and right position based on available space
- [x] Performant API for supporting multiple hovercards in close proximity
- [x] Transitions
- [x] Documentation
  - [x] Website content
  - [x] Website examples
  - [x] Storybook examples
    - [x] In `Card`
    - [x] In `IndexTable.Cell`
      - [x] Customer detail preview
      - [x] Order detail preview

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

### 🎩 checklist

- [ ] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
  - [x] Desktop only feature
- [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [ ] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
  - [x] Ignored by screen readers
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

**Comment by lgriffee** [comment]
Grabbed some stills from your awesome videos for sharing this work out and figured I'd drop them here in case we need them again in the future: 

| Before | After | 
| -- | -- | 
|Screenshot 2023-12-07 at 10 56 44 AM|Screenshot 2023-12-07 at 10 57 20 AM|

File: polaris-react/src/components/PositionedOverlay/utilities/math.ts [hunk 1]
```diff
@@ -90,28 +129,95 @@ export function calculateHorizontalPosition(
   containerRect: Rect,
   overlayMargins: Margins,
   preferredAlignment: PreferredAlignment,
+  _scrollableContainerRect: Rect,
```
Comment (kyledurand) [inline polaris-react/src/components/PositionedOverlay/utilities/math.ts]: Can this be removed?

File: polaris-react/src/components/PositionedOverlay/utilities/math.ts [hunk 2]
```diff
@@ -90,28 +129,95 @@ export function calculateHorizontalPosition(
...
-      maximum,
-      Math.max(0, activatorRight - overlayMargins.horizontal),
-    );
+  const maximumWidth = containerRect.width - overlayRect.width;
+  const activatorRight =
+    containerRect.width - (activatorRect.left + activatorRect.width);
+
+  const minimumSurroundingSpace = overlayMargins.horizontal
+    ? overlayMargins.horizontal
+    : 16;
```
Comment (kyledurand) [inline polaris-react/src/components/PositionedOverlay/utilities/math.ts]: Can this be imported from tokens?

File: polaris-react/src/utilities/index-table/tests/hooks-useRowHovered.test.tsx [hunk 3]
```diff
@@ -7,6 +7,16 @@ import {IndexTable} from '../../../components/IndexTable';
...
+window.matchMedia =
+  window.matchMedia ||
+  function () {
+    return {
+      matches: false,
+      addListener() {},
+      removeListener() {},
+    };
+  };
+
```
Comment (kyledurand) [inline polaris-react/src/utilities/index-table/tests/hooks-useRowHovered.test.tsx]: There are 10 of these. Maybe worth just adding to `setup/tests.ts`? Or `setup/environments.ts`

**Review by kyledurand — COMMENTED** [review]
Looking good! I don't have time for a full review but was curious about the changes to popover so I took a quick look

File: polaris-react/src/components/AlphaHoverCard/AlphaHoverCard.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,680 @@
...
+    };
+
+  const activator = (
+    <Link removeUnderline url="#">
+      Saul Goodman
+    </Link>
+  );
+
+  const toggleActive = (active: boolean) => {
+    console.log(`hover card ${active ? 'active' : 'not active'}`);
```
Comment (laurkim) [inline polaris-react/src/components/AlphaHoverCard/AlphaHoverCard.stories.tsx]: 🔥

File: polaris-react/src/components/AlphaHoverCard/components/HoverCardOverlay.tsx [hunk 5]
```diff
@@ -0,0 +1,154 @@
+import React, {useEffect, useRef} from 'react';
```
Comment (laurkim) [inline polaris-react/src/components/AlphaHoverCard/components/HoverCardOverlay.tsx]:
```suggestion
import React, {useEffect, useRef, useState} from 'react';
```

File: polaris-react/src/components/AlphaHoverCard/components/HoverCardOverlay.tsx [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+  active,
+  zIndexOverride,
+  activator,
+  dynamic,
+  snapToParent,
+  onMouseEnter,
+  onMouseLeave,
+}: HoverCardOverlayProps) {
+  const contentNode = useRef<HTMLDivElement | null>(null);
+  const [shouldAnimate, setShouldAnimate] = React.useState(false);
```
Comment (laurkim) [inline polaris-react/src/components/AlphaHoverCard/components/HoverCardOverlay.tsx]:
```suggestion
  const [shouldAnimate, setShouldAnimate] = useState(false);
```

File: polaris-react/src/components/AlphaHoverCard/components/HoverCardOverlay.tsx [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    positioning,
+    desiredWidth,
+    desiredHeight,
+  }) => {
+    const className = classNames(
+      styles.AlphaHoverCard,
+      snapToParent && styles.snapToParent,
+      positioning === 'above' && styles.positionedAbove,
+      measuring && styles.measuring,
+      !measuring && styles.measured,
```
Comment (laurkim) [inline polaris-react/src/components/AlphaHoverCard/components/HoverCardOverlay.tsx]:
```suggestion
      measuring ? styles.measuring : styles.measured,
```

**Review by laurkim — APPROVED** [review]
This is fantastic!! 🌟💯🥇

Some minor suggestions that aren't blocking. When tophatting Popover stories I did notice some issues with font styles on xs/sm screens but I think that is an unrelated issue that was fixed in a PR merged since this branch has been updated off `main`.

**Comment by heyjoethomas** [comment]
@yurm04 I find the underline on hover confusing. At first I thought it meant the item was also a link I could click to get more details but it just selected the row. I wonder if we could just eliminate the underline all together.

**Comment by yurm04** [comment]
@heyjoethomas could you post a screenshot of which underline you are talking about? I'm still onboarding to this project so i don't know all the different parts yet 😅

**Comment by heyjoethomas** [comment]
@yurm04 '1 Items' is being hovered here and we are applying an underline to the text. Makes it feel like a link.
Screenshot 2024-02-23 at 10 48 35 AM

File: polaris-react/src/components/PositionedOverlay/utilities/math.ts (hunk 1 above)
Comment (yurm04) [reply polaris-react/src/components/PositionedOverlay/utilities/math.ts]: @kyledurand i removed this and the tests started breaking so it's being used somewhere, though i couldn't find where at a quick glance. Going to leave for now and I can dig in more if it bugs us

File: polaris-react/src/components/PositionedOverlay/utilities/math.ts (hunk 2 above)
Comment (yurm04) [reply polaris-react/src/components/PositionedOverlay/utilities/math.ts]: i think our tokens are `rem` strings so it would take some mathing and type casting. Unless we have somewhere in our tokens that's compatible? I couldn't find it but i'm not super familiar with that codebase

**Comment by yurm04** [comment]
@heyjoethomas this seems to be some css clashing with the index table styling because of css specificity. This is a `Link` with `removeUnderline` set, but the index table has some specific styling for links overriding it i believe.

I can change this to a `Text` if you want a specific tone and variant, just let me know. Here's what it looks like as a default `<Text as='p'>` 

Screenshot 2024-02-23 at 11 29 39 AM

**Comment by yurm04** [comment]
just FYI @heyjoethomas that the content inside of the cells is implemented by the user in shopify/web. I can make the change here in our storybook, but every index table will be implemented differently and if this is something we want to roll out elsewhere, we'll have to make those changes in web.

Just letting you know so that you're not confused when this ships and you maybe find links in the admin with underlines.

File: polaris-react/src/components/PositionedOverlay/utilities/math.ts (hunk 2 above)
Comment (kyledurand) [reply polaris-react/src/components/PositionedOverlay/utilities/math.ts]: Yeah maybe, I was thinking we could grab them directly, before they get converted but you'd still have to parseInt to remove `px` so it's still jank

Comment (chloerice) [reply polaris-react/src/components/PositionedOverlay/utilities/math.ts]: The `overlayMargins` arg passed into these functions are grabbed from the overlay's contents as it varies by component (for example Popover has set margins, HoverCard has none). We wouldn't want to use the equivalent space token here because token values/names are subject to change and as you mentioned contain the unit, which `window.getComputedStyle` does not. From main:

https://github.com/Shopify/polaris/blob/14f24f142c0b44f1905859448ae9bc1d9fa6e86e/polaris-react/src/components/PositionedOverlay/PositionedOverlay.tsx#L313-L317

File: polaris-react/src/components/AlphaHoverCard/AlphaHoverCard.stories.tsx [hunk 8]
```diff
@@ -0,0 +1,679 @@
...
+              deliveryMethod,
+              fulfillmentStatus,
+              items,
+            })}
+          >
+            <div style={{minHeight: '100%', padding: 'var(--p-space-150)'}}>
+              <Link monochrome removeUnderline url="#">
+                {`${items.length} items`}
+              </Link>
+            </div>
```
Comment (chloerice) [inline polaris-react/src/components/AlphaHoverCard/AlphaHoverCard.stories.tsx]: This was a copy/paste typo from copying the customers activator, this shouldn't be a link. cc @heyjoethomas 

```suggestion
                {`${items.length} items`}
            </div>
```

**Comment by heyjoethomas** [comment]
@yurm04 Thanks for the context. I think what is here then is 👍 and good to go. No other notes from me.

File: polaris-react/src/components/AlphaHoverCard/AlphaHoverCard.module.scss [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+  display: flex;
+  background: var(--p-color-bg-surface);
+  margin: var(--p-space-100) var(--p-space-200) var(--p-space-400);
+  box-shadow: var(--p-shadow-500);
+  border-radius: var(--p-border-radius-300);
+
+  @include shadow-bevel(
+    $boxShadow: var(--p-shadow-md),
+    $borderRadius: var(--p-border-radius-300),
+    $zIndex: 2
```
Comment (sophschneider) [inline polaris-react/src/components/AlphaHoverCard/AlphaHoverCard.module.scss:19]: I wonder if hover card content could ever be above this 🤔 the card should be in its own portal and have its own stacking context but I could maybe see something being above this. Either way it should be fine for now and we can raise it if needed but just flagging if anyone knows any potential content above `2`

File: polaris-react/src/components/AlphaHoverCard/AlphaHoverCard.module.scss [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+.HoverCardOverlay-enter {
+  opacity: 1;
+}
+
+.HoverCardOverlay-glide {
+  opacity: 1;
+  transition: top var(--p-motion-duration-100) cubic-bezier(0.25, 0, 0.75, 1.35),
+    bottom var(--p-motion-duration-100) cubic-bezier(0.25, 0, 0.75, 1.35);
```
Comment (sophschneider) [inline polaris-react/src/components/AlphaHoverCard/AlphaHoverCard.module.scss:41]: do we need `prefers-reduced-motion` media queries on these animations?

File: polaris-react/src/components/AlphaHoverCard/AlphaHoverCard.module.scss [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    var(--p-motion-duration-300);
+}
+
+.measured {
+  opacity: 1;
+  transition: opacity var(--p-motion-duration-100) var(--p-motion-ease-in-out)
+    var(--p-motion-duration-300);
+}
+
+.positionedAbove {
```
Comment (sophschneider) [inline polaris-react/src/components/AlphaHoverCard/AlphaHoverCard.module.scss:56]: optional: we could capitalize these classes to match the casing for the rest of the file

File: polaris-react/src/components/AlphaHoverCard/components/HoverCardOverlay.tsx [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+      render={renderHoverCard}
+      classNames={className}
+      zIndexOverride={zIndexOverride}
+    />
+  ) : null;
+
+  return overlayMarkup;
+}
+
+export function nodeContainsDescendant(
```
Comment (sophschneider) [inline polaris-react/src/components/AlphaHoverCard/components/HoverCardOverlay.tsx]: is this used anywhere? we can probably delete it

File: polaris-react/src/components/AlphaHoverCard/components/HoverCardOverlay.tsx [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+  const contentNode = useRef<HTMLDivElement | null>(null);
+  const [shouldAnimate, setShouldAnimate] = useState(false);
+
+  useEffect(() => {
+    if (dynamic && active && !shouldAnimate) {
+      setShouldAnimate(true);
+    }
+  }, [dynamic, active, shouldAnimate]);
+
+  const getMinMaxDimensionsOfChildren = () => {
```
Comment (sophschneider) [inline polaris-react/src/components/AlphaHoverCard/components/HoverCardOverlay.tsx]: should this be broken out into its own helper function and take the `contentNode`? Just so its not part of every rerender

File: polaris-react/src/components/AlphaHoverCard/components/HoverCardOverlay.tsx [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+    const className = classNames(
+      styles.AlphaHoverCard,
+      snapToParent && styles.snapToParent,
+      positioning === 'above' && styles.positionedAbove,
+      measuring ? styles.measuring : styles.measured,
+    );
+
+    const dimensions = getMinMaxDimensionsOfChildren();
+    const numericalMaxHeight = dimensions?.maxHeight
+      ? Number(dimensions?.maxHeight.replace(/\D/g, ''))
```
Comment (sophschneider) [inline polaris-react/src/components/AlphaHoverCard/components/HoverCardOverlay.tsx]: could we use something like `parseInt` or `parseFloat` here instead of `Number` and regex? I think it might be safer

File: polaris-react/src/components/PositionedOverlay/utilities/math.ts [hunk 15]
```diff
@@ -143,10 +249,15 @@ export function intersectionWithViewport(
 }
 
 export function windowRect() {
+  const top = window.scrollY;
+  const left = window.scrollX;
+  const height = window.innerHeight;
+  const width = document.body.clientWidth;
+
```
Comment (sophschneider) [inline polaris-react/src/components/PositionedOverlay/utilities/math.ts]: is this change still needed? I don't see the defaults overridden anywhere currently

File: polaris-react/src/components/PositionedOverlay/utilities/math.ts [hunk 16]
```diff
@@ -20,68 +25,102 @@ export function calculateVerticalPosition(
...
-      (distanceToTopScroll >= distanceToBottomScroll &&
-        !enoughSpaceFromBottomScroll)) &&
-      (spaceAbove > desiredHeight || spaceAbove > spaceBelow)
-      ? positionIfAbove
-      : positionIfBelow;
+  let positionIfAbove =
+    activatorTop + containerRectTop - heightIfAbove - verticalMargins;
+  let positionIfBelow = activatorBottom + containerRectTop;
+
+  if (!positionedHorizontal) {
```
Comment (sophschneider) [inline polaris-react/src/components/PositionedOverlay/utilities/math.ts:72]: is this top `if (!positionedHorizontal)` necessary?  below we are already checking if above and if below which are both not left or right

File: polaris-react/src/components/PositionedOverlay/utilities/math.ts [hunk 17]
```diff
@@ -20,68 +25,102 @@ export function calculateVerticalPosition(
...
+  if (enoughSpaceFromTopEdge && enoughSpaceFromBottomEdge) {
+    return {height: heightIfBelow, top: positionIfBelow, positioning: 'below'};
   }
 
-  return distanceToTopScroll > minimumSpaceToScroll
-    ? positionIfAbove
-    : positionIfBelow;
+  return mostSpaceOnBottom
+    ? {height: heightIfBelow, top: positionIfBelow, positioning: 'below'}
+    : {height: heightIfAbove, top: positionIfAbove, positioning: 'above'};
```
Comment (sophschneider) [inline polaris-react/src/components/PositionedOverlay/utilities/math.ts:127]: optional: it could be more readable to put returns these in a switch statement.

This whole function could probably be broken out into some helpers as well to make it a little easier to follow,  but not blocking!

File: polaris-react/src/components/PositionedOverlay/utilities/math.ts (hunk 2 above)
Comment (sophschneider) [reply polaris-react/src/components/PositionedOverlay/utilities/math.ts]: could we just make a global const or something like const `MINIMUM_SURROUNDING_SPACE = 16`? I think its reused on line 32 as well

File: polaris-react/src/components/PositionedOverlay/utilities/math.ts [hunk 18]
```diff
@@ -90,28 +129,95 @@ export function calculateHorizontalPosition(
...
+  const distanceToLeftEdge = activatorRect.left;
+  const distanceToRightEdge = containerRect.width - activatorRect.right;
+  const enoughSpaceFromLeftEdge =
+    distanceToLeftEdge >=
+    (overlayMinWidth || desiredWidth) + minimumSurroundingSpace;
+  const enoughSpaceFromRightEdge =
+    distanceToRightEdge >=
+    (overlayMinWidth || desiredWidth) + minimumSurroundingSpace;
+
+  if (!preferredHorizontalPosition) {
```
Comment (sophschneider) [inline polaris-react/src/components/PositionedOverlay/utilities/math.ts:166]: same as above, we probably don't need this outer if?

File: polaris-react/src/components/IndexTable/IndexTable.tsx [hunk 19]
```diff
@@ -139,25 +141,39 @@ const SCROLL_BAR_DEBOUNCE_PERIOD = 300;
 
 function IndexTableBase({
   headings,
-  bulkActions = [],
+
```
Comment (sophschneider) [inline polaris-react/src/components/IndexTable/IndexTable.tsx]: can we delete this extra space?

File: polaris-react/src/components/PositionedOverlay/utilities/math.ts (hunk 1 above)
Comment (sophschneider) [reply polaris-react/src/components/PositionedOverlay/utilities/math.ts]: it would be best if we refactored these in the future to be object props instead of ordered params in the future

**Review by sophschneider — APPROVED** [review]
Had a few code comments but nothing blocking!!

Tophatted in Chrome, Firefox, and Safari looks great ⭐️

File: polaris-react/src/components/AlphaHoverCard/AlphaHoverCard.module.scss (hunk 10 above)
Comment (yurm04) [reply polaris-react/src/components/AlphaHoverCard/AlphaHoverCard.module.scss:41]: would we want to reduce the animation by some or just remove it altogether for reduced motion? cc @chloerice

File: polaris-react/src/components/AlphaHoverCard/AlphaHoverCard.module.scss (hunk 11 above)
Comment (yurm04) [reply polaris-react/src/components/AlphaHoverCard/AlphaHoverCard.module.scss:56]: it looks like the classes for components are capitalized but `measuring` `measured` and `positionedAbove` are not. I assumed this was because we don't capitalize styles for state-like classnames. Happy to change if that's not the case though

File: polaris-react/src/components/PositionedOverlay/utilities/math.ts (hunk 17 above)
Comment (yurm04) [reply polaris-react/src/components/PositionedOverlay/utilities/math.ts:127]: wasn't sure exactly what you meant @sophschneider so I made an issue and I'm down to pair on it after this ships! https://github.com/Shopify/polaris-internal/issues/1468

File: polaris-react/src/components/PositionedOverlay/utilities/math.ts (hunk 16 above)
Comment (yurm04) [reply polaris-react/src/components/PositionedOverlay/utilities/math.ts:72]: going to leave these for now, i removed them and then things started positioning weirdly so i think they are necessary with the current calcs. But we can dig in after it ships to see if it can be refactored

**Comment by yurm04** [comment]
"I have signed the CLA!"
